lightning/src/routing/router.rs

```diff
@@ -333,79 +494,225 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, targ
+					let known_channel_amount_msat = known_channels_amounts_available_msat.get_mut(&payment_hop.route_hop.short_channel_id).unwrap();
+					let mut used_in_channel_msat = path_bottleneck_msat + payment_hop.total_fee_msat;
+					if idp == payment_path.hops.len() - 1 {
+						// No fees on the last hop.
+						used_in_channel_msat = path_bottleneck_msat;
+					}
+					// This should not overflow because we choose use at most the minimum available channel amount
+					// (a.k.a. path bottleneck) across all used channels in this path.
+					*known_channel_amount_msat -= used_in_channel_msat;
 				}
+				// Track the total amount all our collected paths allow to send so that we:
+				// - know when to stop looking for more paths
+				// - know which of the hops are useless considering how much more sats we need
+				found_amount_msat += payment_path.get_amount_transferred_msat();
+
+				payment_paths.push(payment_path);
+				found_new_path = true;
+				break 'path_construction;
+			}
 
-				let new_entry = match dist.remove(&res.last().unwrap().pubkey) {
-					Some(hop) => hop.3,
-					None => return Err(LightningError{err: "Failed to find a non-fee-overflowing path to the given destination".to_owned(), action: ErrorAction::IgnoreError}),
-				};
-				res.last_mut().unwrap().fee_msat = new_entry.fee_msat;
-				res.last_mut().unwrap().cltv_expiry_delta = new_entry.cltv_expiry_delta;
-				res.push(new_entry);
+			// Otherwise, since the current target node is not us, keep "unrolling" the payment graph from
+			// dest to source by finding a way to reach the current target from the source side.
+			match network.get_nodes().get(&pubkey) {
+				None => {},
+				Some(node) => {
+					add_entries_to_cheapest_to_target_node!(node, &pubkey, lowest_fee_to_node);
+				},
 			}
-			res.last_mut().unwrap().fee_msat = final_value_msat;
-			res.last_mut().unwrap().cltv_expiry_delta = final_cltv;
-			let route = Route { paths: vec![res] };
-			log_trace!(logger, "Got route: {}", log_route!(route));
-			return Ok(route);
 		}
 
-		match network.get_nodes().get(&pubkey) {
-			None => {},
-			Some(node) => {
-				add_entries_to_cheapest_to_target_node!(node, &pubkey, lowest_fee_to_node);
-			},
+		if found_amount_msat >= recommended_available_msat || !found_new_path {
+			break 'paths_collection;
+		}
+	}
+
+	if payment_paths.len() == 0 {
+		return Err(LightningError{err: "Failed to find a path to the given destination".to_owned(), action: ErrorAction::IgnoreError});
+	}
+
+	if found_amount_msat < final_value_msat {
+		return Err(LightningError{err: "Failed to find a sufficient route to the given destination".to_owned(), action: ErrorAction::IgnoreError});
+	}
+
+	// Sort by total fees and take the best paths.
+	payment_paths.sort_by_key(|path| path.get_total_fee_paid_msat());
+	if payment_paths.len() > 50 {
+		payment_paths.truncate(50);
+	}
+
+	// Draw multiple sufficient routes by randomly combining the selected paths.
+	let mut drawn_routes = Vec::new();
```

Comment:
Aaah I finally understand what you aim by scaling by 4 `final_value_msat` to `recommended_available_msat`, thus drawing more paths than necessary from which a random combination of a subset of the found paths will yell the final "logical" path (i.e 1 or more paths)